### PR TITLE
In devMode emulation errors, link to queries to add

### LINKED ISSUE #736: when emulating user in dev mode, create the user if they don't exist
Currently if you try to switch your effective user in the development / docker mode, say to test an allowAccess rule, it gives an "access denied" error. Instead, create the user locally if they don't exist and then switch into them.

### PR DESCRIPTION
**Pull request by trombonekenny**
Closes #736

### CODE REVIEW

File: pages/instructorEffectiveUser/instructorEffectiveUser.js [hunk 1]
```diff
@@ -32,6 +33,18 @@ router.post('/', function(req, res, next) {
         res.clearCookie('pl_requested_mode');
         res.redirect(req.originalUrl);
     } else if (req.body.__action == 'changeUid') {
+        if (config.devMode) {
+            // Add the user in devmode to test access rules
+            var params = {
+                uid: req.body.pl_requested_uid,
+                course_instance_id: res.locals.course_instance.id,
+            };
+            return sqldb.query(sql.add_and_enroll, params, function(err) {
```
Comment (mwest1066) [inline pages/instructorEffectiveUser/instructorEffectiveUser.js]: Looks like a race condition here -- we should do the `res.cookie()` and `res.redirect()` inside the callback (which will require duplicating them, I know).

Comment (trombonekenny) [reply pages/instructorEffectiveUser/instructorEffectiveUser.js]: Doesn't it already do that (see lines 44, 45)? I don't think I'm following. 

The `return` before sqldb.query ends that branch after the callback is run so lines 48-49 never get run. That's why they're duplicated.

Comment (mwest1066) [reply pages/instructorEffectiveUser/instructorEffectiveUser.js]: You are entirely correct, I was mis-reading the code. Sorry! I'd prefer to change it to remove the `return` on line 42 and instead put 48,49 inside an `else` block, which I find makes the control-flow easier to understand (all paths terminate exactly once in their terminal blocks).

File: pages/instructorEffectiveUser/instructorEffectiveUser.sql [hunk 2]
```diff
@@ -13,3 +13,14 @@ SELECT
...
+-- BLOCK add_and_enroll
+WITH rows as (
+    INSERT INTO users (uid, name) VALUES ($uid, $uid)
+    ON CONFLICT (uid) DO UPDATE SET uid=$uid /* needed for the return */
+    RETURNING user_id
+)
+INSERT INTO enrollments (user_id, course_instance_id, role)
+    SELECT user_id, $course_instance_id, 'Student'
+    FROM rows
+ON CONFLICT DO NOTHING;
```
Comment (tbretl) [inline pages/instructorEffectiveUser/instructorEffectiveUser.sql]: This code adds a user with the given UID if that user does not exist, and enrolls this new user in the given course instance.

A corner case is when this user already exists, but is not enrolled in the given course instance. In this case, you still get "access denied" (it triggers the `ON CONFLICT`). It seems like the right behavior would be:

- If the user does not exist, create one, and enroll in the course instance as student.
- If the user does exist and is not enrolled in the course instance, enroll in the course instance as student.
- If the user does exist and is enrolled in the course instance, do nothing.

Does that make sense?

File: pages/instructorEffectiveUser/instructorEffectiveUser.js [hunk 3]
```diff
@@ -32,8 +33,21 @@ router.post('/', function(req, res, next) {
         res.clearCookie('pl_requested_mode');
         res.redirect(req.originalUrl);
     } else if (req.body.__action == 'changeUid') {
-        res.cookie('pl_requested_uid', req.body.pl_requested_uid, {maxAge: 60 * 60 * 1000});
-        res.redirect(req.originalUrl);
+        if (config.devMode) {
+            // Add the user in devmode to test access rules
```
Comment (tbretl) [inline pages/instructorEffectiveUser/instructorEffectiveUser.js]: May want to be more explicit here - for example, make clear that nothing will happen if the user already exists.

**Comment by tbretl** [comment]
One more quick note - if you switch the effective user in one course instance, then switch to a different course instance, you again get "access denied". This may be desired behavior (although the error message is pretty cryptic).

**Comment by mwest1066** [comment]
@trombonekenny bumping this PR as needing attention

**Comment by trombonekenny** [comment]
I think that last push resolves Tim's concerns and is ready for (re-)review.

**Comment by trombonekenny** [comment]
Reviewing this old code, and the state of the permission and UX changes since this started, I tried a few things and settled on the easiest - in an error message about the user not existing, when in devMode, link to the admin query that can add the user. I think that resolves the root issue, doesn't change the (now existing) code or test base, and is a principle of least surprise because the devMode admin needs to manually take the action they want.

**Review by mwest1066 — APPROVED** [review]

**Pull request merged**